ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,570 @@
+	}
+	db.compTimeMeter = metrics.NewRegisteredMeter(namespace+"compact/time", nil)
+	db.compReadMeter = metrics.NewRegisteredMeter(namespace+"compact/input", nil)
+	db.compWriteMeter = metrics.NewRegisteredMeter(namespace+"compact/output", nil)
+	db.diskSizeGauge = metrics.NewRegisteredGauge(namespace+"disk/size", nil)
+	db.diskReadMeter = metrics.NewRegisteredMeter(namespace+"disk/read", nil)
+	db.diskWriteMeter = metrics.NewRegisteredMeter(namespace+"disk/write", nil)
+	db.writeDelayMeter = metrics.NewRegisteredMeter(namespace+"compact/writedelay/duration", nil)
+	db.writeDelayNMeter = metrics.NewRegisteredMeter(namespace+"compact/writedelay/counter", nil)
+	db.memCompGauge = metrics.NewRegisteredGauge(namespace+"compact/memory", nil)
+	db.level0CompGauge = metrics.NewRegisteredGauge(namespace+"compact/level0", nil)
+	db.nonlevel0CompGauge = metrics.NewRegisteredGauge(namespace+"compact/nonlevel0", nil)
+	db.seekCompGauge = metrics.NewRegisteredGauge(namespace+"compact/seek", nil)
+	db.manualMemAllocGauge = metrics.NewRegisteredGauge(namespace+"memory/manualalloc", nil)
+
+	// Start up the metrics gathering and return
+	go db.meter(metricsGatheringInterval)
+	return db, nil
+}
+
+// Close stops the metrics collection, flushes any pending data to disk and closes
+// all io accesses to the underlying key-value store.
+func (d *Database) Close() error {
+	d.quitLock.Lock()
+	defer d.quitLock.Unlock()
+
+	if d.quitChan != nil {
+		errc := make(chan error)
+		d.quitChan <- errc
+		if err := <-errc; err != nil {
+			d.log.Error("Metrics collection failed", "err", err)
+		}
+		d.quitChan = nil
+	}
+	return d.db.Close()
+}
+
+// Has retrieves if a key is present in the key-value store.
+func (d *Database) Has(key []byte) (bool, error) {
+	_, closer, err := d.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	closer.Close()
+	return true, nil
+}
+
+// Get retrieves the given key if it's present in the key-value store.
+func (d *Database) Get(key []byte) ([]byte, error) {
+	dat, closer, err := d.db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]byte, len(dat))
+	copy(ret, dat)
+	closer.Close()
+	return ret, nil
+}
+
+// Put inserts the given value into the key-value store.
+func (d *Database) Put(key []byte, value []byte) error {
+	return d.db.Set(key, value, pebble.NoSync)
+}
+
+// Delete removes the key from the key-value store.
+func (d *Database) Delete(key []byte) error {
+	return d.db.Delete(key, nil)
+}
+
+// NewBatch creates a write-only key-value store that buffers changes to its host
+// database until a final write is called.
+func (d *Database) NewBatch() ethdb.Batch {
+	return &batch{
+		b: d.db.NewBatch(),
+	}
+}
+
+// NewBatchWithSize creates a write-only database batch with pre-allocated buffer.
+// TODO can't do this with pebble.  Batches are allocated in a pool so maybe this doesn't matter?
+func (d *Database) NewBatchWithSize(_ int) ethdb.Batch {
```

Comment:
I need to investigate it a bit. Batch with pre-allocated internal buffer is very useful for PBSS. 
Pebble has `LargeBatch` feature not sure what it is, but anyway I think it's even not hard to implement it in pebble.
Nothing to complain, just a reminder for myself.